File: frontend/src/lib/chapter-length.ts
export const WORD_COUNT_TARGET_MAX = 100_000;

export type ChapterPace = "none" | "short" | "met" | "over";

export interface ChapterLengthProgress {
  pace: ChapterPace;
  written: number;
  target: number | null;
  remaining: number;
  over: number;
}

export function readWordCountTarget(value: unknown): number | null {
  if (typeof value !== "number" || !Number.isInteger(value)) return null;
  if (value < 1 || value > WORD_COUNT_TARGET_MAX) return null;
  return value;
}

export function parseWordCountTarget(
  raw: string,
): { ok: true; value: number | null } | { ok: false } {
  const trimmed = raw.trim();
  if (trimmed === "") return { ok: true, value: null };
  if (!/^\d+$/.test(trimmed)) return { ok: false };
  const value = Number(trimmed);
  if (!Number.isSafeInteger(value) || value < 1 || value > WORD_COUNT_TARGET_MAX) {
    return { ok: false };
  }
  return { ok: true, value };
}

export function chapterLengthProgress(
  written: number,
  target: number | null,
): ChapterLengthProgress {
  if (target == null) {
    return { pace: "none", written, target: null, remaining: 0, over: 0 };
  }
  if (written < target) {
    return { pace: "short", written, target, remaining: target - written, over: 0 };
  }
  if (written > target) {
    return { pace: "over", written, target, remaining: 0, over: written - target };
  }
  return { pace: "met", written, target, remaining: 0, over: 0 };
}

type LengthTranslator = (key: string, options?: { count: number }) => string;

export interface CorkboardLengthLabel {
  pace: Exclude<ChapterPace, "none">;
  text: string;
}

/**
 * 软木板卡片上的字数差距。没有目标时不显示。
 * 刚好达到标「达标」，不写成还差 0。超出用超出，不用还差。
 */
export function corkboardLengthLabel(
  written: number,
  target: number | null,
  translate: LengthTranslator,
): CorkboardLengthLabel | null {
  const progress = chapterLengthProgress(written, target);
  if (progress.pace === "none") return null;
  if (progress.pace === "short") {
    return {
      pace: "short",
      text: translate("writing.chapterLength.remaining", { count: progress.remaining }),
    };
  }
  if (progress.pace === "over") {
    return {
      pace: "over",
      text: translate("writing.chapterLength.over", { count: progress.over }),
    };
  }
  return { pace: "met", text: translate("writing.chapterLength.onTarget") };
}

File: backend/app/storage/chapter_length.py
# -*- coding: utf-8 -*-
"""章节字数目标。

目标用来看这一章写短了还是拖了。空值表示不设目标。
不记录写作活动，也不改正文。
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

# 一章的目标再长也不该超过一部中篇。超过这个数对节奏没有意义。
WORD_COUNT_TARGET_MAX = 100_000

ChapterPace = Literal["none", "short", "met", "over"]


def normalize_word_count_target(value: int | None) -> int | None:
    """校验章节字数目标。None 表示清空。"""
    if value is None:
        return None
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValueError("章节字数目标必须是整数")
    if value < 1 or value > WORD_COUNT_TARGET_MAX:
        raise ValueError(f"章节字数目标需在 1 到 {WORD_COUNT_TARGET_MAX} 之间")
    return value


def read_word_count_target(chapter: object) -> int | None:
    """读取已保存的目标。缺省或非法值视为未设置。"""
    raw = getattr(chapter, "word_count_target", None)
    if isinstance(raw, bool) or not isinstance(raw, int):
        return None
    if raw < 1 or raw > WORD_COUNT_TARGET_MAX:
        return None
    return raw


@dataclass(frozen=True)
class ChapterLengthProgress:
    """已写字数相对目标的差距。written 必须来自现有字数统计。"""

    pace: ChapterPace
    written: int
    target: int | None
    remaining: int
    over: int


def chapter_length_progress(written: int, target: int | None) -> ChapterLengthProgress:
    """比较已写字数和目标。没有目标时不判断长短。"""
    if target is None:
        return ChapterLengthProgress(
            pace="none",
            written=written,
            target=None,
            remaining=0,
            over=0,
        )
    if written < target:
        return ChapterLengthProgress(
            pace="short",
            written=written,
            target=target,
            remaining=target - written,
            over=0,
        )
    if written > target:
        return ChapterLengthProgress(
            pace="over",
            written=written,
            target=target,
            remaining=0,
            over=written - target,
        )
    return ChapterLengthProgress(
        pace="met",
        written=written,
        target=target,
        remaining=0,
        over=0,
    )
